usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -11984,3 +12019,87 @@ static void put_target_info(STDLL_TokData_t *tokdata,
     }
 }
 
+/*
+ * Must be called either with WRITE_LOCK on the object that owns the template
+ * specified as @tmpl, or before the object is made publicly available via
+ * object_mgr_create_final.
+ */
+static CK_RV update_ep11_attrs_from_blob(STDLL_TokData_t *tokdata,
+                                         TEMPLATE *tmpl)
+{
+    ep11_private_data_t *ep11_data = tokdata->private_data;
+    CK_BBOOL restr = CK_FALSE; /*, never_mod = CK_FALSE; */
+    CK_BBOOL attrb = CK_FALSE, useasdata = CK_FALSE;
+    CK_BBOOL pkeyextr = CK_FALSE, pkeyneverextr = CK_FALSE;
+    CK_ULONG stdcomp1 = 0;
+    CK_ATTRIBUTE *attr, *blob_attr = NULL;
+    ep11_target_info_t* target_info;
+    CK_RV rc = CKR_OK;
+    CK_ULONG i;
+
+    CK_ATTRIBUTE ibm_attrs[] = {
+        { CKA_IBM_RESTRICTABLE, &restr, sizeof(restr) },
+     /* Skip CKA_IBM_NEVER_MODIFIABLE for now, it causes CKR_ARGUMENTS_BAD
+        { CKA_IBM_NEVER_MODIFIABLE, &never_mod, sizeof(never_mod) }, */
+        { CKA_IBM_USE_AS_DATA, &useasdata, sizeof(useasdata) },
+        { CKA_IBM_ATTRBOUND, &attrb, sizeof(attrb) },
```

Comment:
With this change, `CKA_IBM_ATTRBOUND` is attached to every EP11 key.  This makes some `if` statements in the AB implementation trivial.  I did not add the attribute everywhere since attribute lookup in OCK is linear in the number of elements and adding an element is prepending (i.e., the commonly used attributes like key type, modulus, prime, ... are moved even further away).

Reply:
Without adding these attribute to the OCK object, an application can't retrieve them (unless they were specified at object creation in the template). Since the EP11 key blob does contain these attributes in any case, I think they should also be visible in the OCK object in any case. 

Yas this increases the linear search time, but I think searching in about 20 attributes does not really cause much trouble.